clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -1209,9 +1227,18 @@ public void handle(OffsetCommitResponse commitResponse, RequestFuture<Void> futu
                             return;
                         } else if (error == Errors.UNKNOWN_MEMBER_ID
                                 || error == Errors.ILLEGAL_GENERATION) {
-                            // need to reset generation and re-join group
-                            resetGenerationOnResponseError(ApiKeys.OFFSET_COMMIT, error);
-                            future.raise(new CommitFailedException());
+                            log.info("OffsetCommit failed with {}: {}", sentGeneration, error.message());
+
+                            // only need to reset generation and re-join group if generation has not changed;
+                            // otherwise only raise rebalance-in-progress error
+                            if (generationUnchanged()) {
+                                resetGenerationOnResponseError(ApiKeys.OFFSET_COMMIT, error);
+                                future.raise(new CommitFailedException());
+                            } else {
+                                future.raise(new RebalanceInProgressException("Offset commit cannot be completed since the " +
```

Comment:
This seems a bit misleading, because the consumer is not actually participating in an ongoing rebalance (yet)

Reply:
@guozhangwang can you elaborate on why we don't just throw `CommitFailedException` here? It seems like it must be the case that the caller is a zombie and should get that exception to alert it, not the `RebalanceInProgressException` which seems to indicate it might be recoverable

Reply:
If the generation has changed since the commit request is sent, then it is likely that it has participated in a new rebalance (and hence get a new generation), or it has reset its generation due to the heartbeat failure. So this commit failure is recoverable.

Reply:
Chatted about this offline, will just leave the concrete proposal here:
```
...
} else {
    if (state == MemberState.REBALANCING) {
        future.raise(new RebalanceInProgressException(...)
    } else {
        future.raise(new CommitFailedException(...)
    }
}
```

Reply:
The reasoning being that we use `CommitFailedException` to signal we have dropped out of the group, and `RebalanceInProgressException` to signal that a rebalance is in progress. There are two cases to consider (within the general case of the generation having changed):
1.  If the generation is unknown and the state is `STABLE` this means we have dropped out of the group, but haven't yet rejoined and haven't invoked `onPartitionsLost` --> should throw `CommitFailed`
2. If the generation is unknown and the state is `REBALANCING` this means we dropped out of the group, but have already noticed and rejoined, and already invoked `onPartitionsLost` --> should throw `RebalanceInProgress`

Note that if we dropped out of the group and already _completed_ the rejoin, the state will be `STABLE` again but the generation will also have been set so this case does not apply. Basically, we want to keep `CommitFailedException` to indicate that the consumer definitely dropped out and will have to rejoin, which is the case in 1. above